File: LeetCode/Liang/221MaximalSquare.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if (matrix.empty()) {
            return 0;
        }
        
        int maxLen = 0;
        vector<vector<int>> DP(matrix.size(), vector<int>(matrix.at(0).size(), 0));
        for (int i = 0; i < matrix.size(); ++i) {
            for (int j = 0; j < matrix.at(0).size(); ++j) {
                if (matrix.at(i).at(j) == '1') {
                    if (i == 0 || j == 0) {
                        DP.at(i).at(j) = 1;
                    } else {
                        DP.at(i).at(j) = min(DP.at(i).at(j - 1), min(DP.at(i - 1).at(j), DP.at(i - 1).at(j - 1))) + 1;
                    }
                    maxLen = max(maxLen, DP.at(i).at(j)); 
                }
            }
        }
        return maxLen * maxLen;
    }
};

File: LeetCode/Hao/221MaximalSquare.java
public class Solution {
    public int maximalSquare(char[][] matrix) {
        int rowMax = matrix.length;
        if (rowMax == 0) {
            return 0;
        }
        int colMax = matrix[0].length;
        int[][] d = new int[rowMax][colMax];
        int max = 0;        
        
        for (int i = 0; i < rowMax; i ++) {
            for (int j = 0; j < colMax; j ++) {
                d[i][j] = matrix[i][j] == '1' ? 1 : 0;
            
                max = Math.max(max, d[i][j]);
            }
        }
        
        for (int i = 1; i < rowMax; i ++) {
            for (int j = 1; j < colMax; j ++) {
                if (matrix[i][j] == '1') {
                    d[i][j] = Math.min(d[i - 1][j - 1], Math.min(d[i - 1][j], d[i][j - 1])) + 1;
                } 
                max = Math.max(max, d[i][j]);
            }
        }
        
        return max * max;
    }
}
